linera-execution/src/wasm/revm.rs

```diff
@@ -0,0 +1,439 @@
+        Ok(())
+    }
+}
+
+fn process_execution_result(result: ExecutionResult) -> Result<(Output, Vec<Log>), ExecutionError> {
+    match result {
+        ExecutionResult::Success {
+            reason: _,
+            gas_used: _,
+            gas_refunded: _,
+            logs,
+            output,
+        } => Ok((output, logs)),
+        ExecutionResult::Revert { gas_used, output } => {
+            let error = RevmExecutionError::Revert { gas_used, output };
+            let error = VmExecutionError::RevmExecutionError(error);
+            Err(ExecutionError::VmError(error))
+        }
+        ExecutionResult::Halt { gas_used, reason } => {
+            let error = RevmExecutionError::Halt { gas_used, reason };
+            let error = VmExecutionError::RevmExecutionError(error);
+            Err(ExecutionError::VmError(error))
+        }
+    }
+}
+
+impl<Runtime> RevmContractInstance<Runtime>
+where
+    Runtime: ContractRuntime,
+{
+    pub fn prepare(module: Vec<u8>, runtime: Runtime) -> Self {
+        let db = DatabaseRuntime::new(runtime);
+        Self { module, db }
+    }
+
+    fn transact_commit_tx_data(
+        &mut self,
+        ch: Choice,
+        tx_data: Bytes,
+    ) -> Result<(Output, Vec<Log>), ExecutionError> {
+        let kind = match ch {
+            Choice::Create => TxKind::Create,
+            Choice::Call => TxKind::Call(self.db.contract_address),
+        };
+        let mut evm: Evm<'_, (), _> = Evm::builder()
+            .with_ref_db(&mut self.db)
+            .modify_tx_env(|tx| {
+                tx.clear();
+                tx.transact_to = kind;
+                tx.data = tx_data;
+            })
+            .build();
+
+        let result = evm.transact_commit();
+        let result = match result {
+            Ok(result) => result,
+            Err(error) => {
+                let error = format!("{:?}", error);
+                let error = RevmExecutionError::TransactCommitError(error);
+                let error = VmExecutionError::RevmExecutionError(error);
+                return Err(ExecutionError::VmError(error));
+            }
+        };
+        process_execution_result(result)
+    }
+
+    fn write_logs(
+        &mut self,
+        contract_address: &Address,
+        logs: Vec<Log>,
+    ) -> Result<(), ExecutionError> {
+        if !logs.is_empty() {
+            let mut runtime = self.db.runtime.lock().expect("The lock should be possible");
+            let stream_name = bcs::to_bytes("ethereum_event").unwrap();
+            let stream_name = StreamName(stream_name);
+            for (log, index) in logs.iter().enumerate() {
+                let mut key = bcs::to_bytes(&contract_address).unwrap();
+                bcs::serialize_into(&mut key, "deploy").unwrap();
+                bcs::serialize_into(&mut key, index).unwrap();
+                let value = bcs::to_bytes(&log).unwrap();
+                runtime.emit(stream_name.clone(), key, value)?;
```

Comment:
Nice!

But I wonder whether it would be better to make the key only the index, and make the contract address part of the stream name?

Also, we'll probably need to persist the current index, so that the events of the next block don't overwrite the previous ones. (We'll disallow that anyway.)

Why do we need the "`deploy`" in the key?

I'm also wondering whether we need to change our own event structure, and make sure we always have an incrementing number as a key.

Reply:
Here the contract address is always the same, so that is a little premature.
Maybe I should have put that for next PR.
But at least, here the Ethereum/Linera world are matching fairly reasonably.